services/datalake/pod-datalake/src/datalake/db.ts

```diff
@@ -0,0 +1,461 @@
+//
+// Copyright © 2025 Hardcore Engineering Inc.
+//
+// Licensed under the Eclipse Public License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may
+// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import { MeasureContext } from '@hcengineering/core'
+import postgres, { type Row } from 'postgres'
+import { type Location, type UUID } from './types'
+import { type RetryOptions, retry } from './retry'
+
+export interface BlobDataId {
+  hash: UUID
+  location: Location
+}
+
+export interface BlobDataRecord extends BlobDataId {
+  filename: string
+  size: number
+  type: string
+}
+
+export interface BlobId {
+  workspace: string
+  name: string
+}
+export interface BlobIds {
+  workspace: string
+  names: string[]
+}
+
+export interface BlobRecord extends BlobId {
+  hash: UUID
+  location: Location
+  parent?: string
+}
+
+export type BlobWithDataRecord = BlobRecord & BlobDataRecord
+
+export interface ListBlobResult {
+  cursor: string | undefined
+  blobs: BlobWithDataRecord[]
+}
+
+export interface StatsResult {
+  workspaces: number
+  blob: {
+    count: number
+    size: number
+  }
+  data: {
+    count: number
+    size: number
+  }
+}
+export interface WorkspaceStatsResult {
+  count: number
+  size: number
+}
+
+export function createDb (ctx: MeasureContext, connectionString: string): BlobDB {
+  const sql = postgres(connectionString, {
+    max: 5,
+    connection: {
+      application_name: 'datalake'
+    },
+    fetch_types: false,
+    prepare: false,
+    types: {
+      // https://jdbc.postgresql.org/documentation/publicapi/constant-values.html
+      int8: {
+        to: 0,
+        from: [20],
+        serialize: (value: string) => value.toString(),
+        parse: (value: number) => Number(value)
+      }
+    }
+  })
+
+  const dbAdapter = new PostgresDBAdapter(sql)
+  return new LoggedDB(new PostgresDB(new RetryDBAdapter(dbAdapter, { retries: 5 })), ctx)
+}
+
+export async function withPostgres<T> (
+  ctx: MeasureContext,
+  connectionString: string,
+  fn: (db: BlobDB) => Promise<T>
+): Promise<T> {
+  const sql = postgres(connectionString, {
```

Comment:
Same stuff two times a row?

Reply:
unused from worker, will double check and remove